orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/Stage.java

```diff
@@ -402,6 +403,106 @@ public Task taskById(String taskId) {
     }
   }
 
+  /**
+   * Computes all ancestor stages, including those of parent pipelines
+   *
+   * @return list of ancestor stages
+   */
+  public List<Stage> ancestorsWithParentPipelines() {
+    return Stage.getAncestors(this, this.execution);
+  }
+
+  private static List<Stage> getAncestors(Stage stage, Execution execution) {
+    List<Stage> allPriorStages = new ArrayList<>();
+
+    if (stage != null && !stage.getRequisiteStageRefIds().isEmpty()) {
+      List<Stage> previousStages =
+          execution.getStages().stream()
+              .filter(s -> stage.getRequisiteStageRefIds().contains(s.getRefId()))
+              .collect(toList());
+
+      Set<String> previousStageIds =
+          new HashSet<>(previousStages.stream().map(Stage::getId).collect(toList()));
+      List<Stage> syntheticStages =
+          execution.getStages().stream()
+              .filter(s -> previousStageIds.contains(s.getParentStageId()))
+              .collect(toList());
+
+      allPriorStages.addAll(previousStages);
+      allPriorStages.addAll(syntheticStages);
+
+      for (Stage s : previousStages) {
+        allPriorStages.addAll(getAncestors(s, execution));
+      }
+    } else if ((stage != null) && !Strings.isNullOrEmpty(stage.getParentStageId())) {
+      Optional<Stage> parent =
+          execution.getStages().stream()
+              .filter(s -> s.getId().equals(stage.getParentStageId()))
+              .findFirst();
+
+      if (!parent.isPresent()) {
+        throw new IllegalStateException(
+            "Couldn't find parent of stage "
+                + stage.getId()
+                + " with parent "
+                + stage.getParentStageId());
+      }
+
+      allPriorStages.add(parent.get());
+      allPriorStages.addAll(getAncestors(parent.get(), execution));
+    } else if ((execution.getType() == PIPELINE)
+        && (execution.getTrigger() instanceof PipelineTrigger)) {
+      PipelineTrigger parentTrigger = (PipelineTrigger) execution.getTrigger();
+
+      Execution parentPipelineExecution = parentTrigger.getParentExecution();
+      String parentPipelineStageId = parentTrigger.getParentPipelineStageId();
+
+      Optional<Stage> parentPipelineStage =
+          parentPipelineExecution.getStages().stream()
+              .filter(
+                  s -> s.getType().equals("pipeline") && s.getId().equals(parentPipelineStageId))
+              .findFirst();
+
+      if (parentPipelineStage.isPresent()) {
+        allPriorStages.addAll(getAncestors(parentPipelineStage.get(), parentPipelineExecution));
+      } else {
+        List<Stage> parentPipelineStages = new ArrayList<>(parentPipelineExecution.getStages());
+        parentPipelineStages.sort(
+            new Comparator<Stage>() {
+              @Override
+              public int compare(Stage s1, Stage s2) {
```

Comment:
i tried doing `Comparator.comparingNullLast(Comparator.comparing(stage::getEndTime))` but it was throwing NPE, if you know what I was doing wrong, would love to simplify this to use an off the shelf comparator

Reply:
do we have any context on why this time comparison is required ?

Reply:
this is a great mystery... hence my comment about this being a "port to java". I wonder if it's there for unit tests?
Or, more likely, this is a relic of the past that could be removed? @ajordens / @duftler do either of you happen to know the history here? (https://github.com/spinnaker/orca/pull/874)